homeassistant/components/solax/config_flow.py

```diff
@@ -5,37 +5,59 @@
 import logging
 from typing import Any
 
-from solax import real_time_api
-from solax.discovery import DiscoveryError
+from solax import discover
+from solax.discovery import REGISTRY, DiscoveryError
 import voluptuous as vol
 
 from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
 from homeassistant.const import CONF_IP_ADDRESS, CONF_PASSWORD, CONF_PORT
+from homeassistant.helpers import selector
 import homeassistant.helpers.config_validation as cv
 
-from .const import DOMAIN
+from .const import CONF_SOLAX_INVERTER, DOMAIN
 
 _LOGGER = logging.getLogger(__name__)
 
 DEFAULT_PORT = 80
 DEFAULT_PASSWORD = ""
+DEFAULT_INVERTER = ""
+
+REGISTRY_HASH = {cls.__name__: cls for cls in REGISTRY}
 
 STEP_USER_DATA_SCHEMA = vol.Schema(
     {
         vol.Required(CONF_IP_ADDRESS): cv.string,
         vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
         vol.Optional(CONF_PASSWORD, default=DEFAULT_PASSWORD): cv.string,
+        vol.Optional(
+            CONF_SOLAX_INVERTER, default=DEFAULT_INVERTER
+        ): selector.SelectSelector(
+            selector.SelectSelectorConfig(
+                options=list(REGISTRY_HASH.keys()),
+                mode=selector.SelectSelectorMode.DROPDOWN,
+                multiple=True,
+            )
+        ),
     }
 )
 
 
 async def validate_api(data) -> str:
     """Validate the credentials."""
 
-    api = await real_time_api(
-        data[CONF_IP_ADDRESS], data[CONF_PORT], data[CONF_PASSWORD]
+    _LOGGER.debug("CONF_SOLAX_INVERTER entry: %s", data[CONF_SOLAX_INVERTER])
+
+    invset = set()
+    for cls_name in data[CONF_SOLAX_INVERTER]:
+        invset.add(REGISTRY_HASH[cls_name])
+    inverter = await discover(
+        data[CONF_IP_ADDRESS],
+        data[CONF_PORT],
+        data[CONF_PASSWORD],
+        inverters=invset,
     )
```

Comment:
I should have fixed the main cause behind the randon crashing during discovery in 3.1.0. But since [X1 Mini V34 can also handle the response from an X1 Boost](https://github.com/squishykid/solax/pull/119#issuecomment-1646691601) it would, in my opinion, be best to detect such cases and have the user deselect X1 Mini V34 in this example.

Reply:
I forgot to mention I only want to use `ALL_COMPLETED` only during initial configuration / [reconfiguration](https://developers.home-assistant.io/blog/2024/03/21/config-entry-reconfigure-step/). In `async_setup_entry()` I want to use `FIRST_COMPLETED` and pass it a single Inverter class.

Reply:
ALL_COMPLETED does not work on my installation, I get timeouts with the following code I use to test the upstream solax library:
`    inverters = loop.run_until_complete(solax.discover("solax.home.arpa", 80, "XXXXXXXXXX", return_when=asyncio.ALL_COMPLETED))
`
I think my inverter (X1HybridG4) crashes in this case. With FIRST_COMPLETED it works.